### DISCUSSION TITLE
Improve Cross-environment portability of libp2p core

### DISCUSSION THREAD

**Discussion by maschad** [libp2p/js-libp2p#2138] [Ideas]
https://github.com/libp2p/js-libp2p/issues/2110 raised an issue about the portability of js-libp2p's core modules, js bundlers are going away anytime soon, and so consumers may run into these issues again depending on their configurations. As @wemeetagain [suggested](https://github.com/libp2p/js-libp2p/pull/2134#issuecomment-1751122412) to avoid the hassle of bundlers, whilst making libp2p truly portable, we should: 

a) stay away from modules which are not cross-environment.

b) if we do need to use a module that is environment specific, we should at least conditionally import it and/or return a no-op for the inapplicable environment.

**Comment by wemeetagain** [comment]
Would like to add a specific actionable result or subissue:
- ability to consume libp2p from browser without a bundler

**Reply by achingbrain** [reply]
I'd like this too, however it would require removing all CJS code from our stack - there are still a whole bunch of dependencies we have that are not ESM.

**Reply by wemeetagain** [reply]
we should list and kill one by one, there can't be too many left

**Reply by achingbrain** [reply]
It's a bunch of the older modules - `node-forge` springs instantly to mind, also `multicast-dns` etc

**Comment by wemeetagain** [comment]
Copying my comment from #2134 

---
I would like to see js-libp2p not require a bundler at all in order to consume in the browser. Related to our portability initiative.

I think that js-libp2p core should be as portable as possible, and that means using isomorphic, or at least portable code throughout.

Also this case, specifically, feels a bit silly. We're requiring browser consumers to bundle an events polyfill for a nodejs-specific feature (using setMaxListeners to avoid a warning) with no benefit to the browser consumer.

**Reply by achingbrain** [reply]
Completely agree with the above.

Since we smother any errors encountered when increasing the number of listeners, we could have solved this by adding `"events": false` to the `"browser"` field in `package.json`, but then we have to put up with lots of warnings on build:

```
▲ [WARNING] Import "setMaxListeners" will always be undefined because the file "(disabled):../../node_modules/events/events.js" has no exports [import-is-undefined]

    dist/src/connection-manager/utils.js:1:9:
      1 │ import { setMaxListeners } from 'events';
````

`"events": {}` instead yields one warning of:

```
▲ [WARNING] Each "browser" mapping must be a string or a boolean [package.json]

    package.json:201:14:
      201 │     "events": {}
          ╵               ^
```

..which sounds bad but the tests pass 🤷 

Though previously we wanted to get rid of the `try/catch` because `setMaxListeners` is in all supported node versions now but we can't have both.

**Reply by wemeetagain** [reply]
Can we move the complexity to `@libp2p/interface` / `@libp2p/interface/events`?

eg:
```ts
// in packages/interface/package.json

"browser": { "events": {} },
```

```ts
// in packages/interface/src/events.ts

import { setMaxListeners as nodeSetMaxListeners } from 'events'
export function setMaxListeners(...) {
  try {
    nodeSetMaxListeners(...)
  } catch (e) {
    // swallow errors, gulp
  }
}
```

```ts
// in packages/libp2p and beyond

import {setMaxListeners} from '@libp2p/interface/events'

...
setMaxListeners(...)
```

**Reply by achingbrain** [reply]
Yes, that'd be an improvement.

I think we should also verify where this is still necessary.  The classic one was while dialling - if a node had more than 10x addresses we'd have an abort signal with more than 10x listeners.  Since smart-dialling we only dial one address at a time now so we might not trigger the warning.

The DHT increases the number of listeners on the query abort signals but the queries involve a lot of dialling, but we've calmed the dialling down so it may be removable too.
